Enable Declarative Validation for ClusterRole

itzPranshul opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

-Enables validation-gen for rbac v1/v1alpha1/v1beta1 groups: `+k8s:validation-gen` tags in `pkg/apis/rbac/(v1,v1alpha1,v1beta1)/doc.go` enable generation for the ClusterRole API group.

- Updates ClusterRole strategy to run generated declarative validation in Validate and ValidateUpdate methods.


#### Which issue(s) this PR is related to:

part of #134280

#### Special notes for your reviewer:

This PR does not migrate any validations: validation rules migration would be introduced in separate PRs.

#### Does this PR introduce a user-facing change?

```release-note
This PR only enables the declarative validation framework for ClusterRole.
Existing validation behaviour remains unchanged unless feature gates DeclarativeValidation or DeclarativeValidationTakeover are explicitly enabled.
```

#### Additional documentation, e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md
```

lengrongfu committed:
feat: add resourcequota WarningsOnCreate request less than limits

Signed-off-by: rongfu.leng <lenronfu@gmail.com>

aojea committed:
ServiceCIDR ValidationAdmissionPolicy for backkwards compatible behavior

Add an integration test with a ValidationAdmissionPolicy that allows to
keep the previous behavior to the introduction of the ServiceCIDR
feature.

Change-Id: I530d06c111f480e8ec535b250d651beb6f0249bc

thockin committed:
Eliminate public ValidateReplicationControllerName

Everyone who referenced it now uses the underlying function.  Clearer
and frees me up to change objectMeta validation without impacting anyone
else.

thockin committed:
Validate ReplicationController.metadata.name

This relies on `+k8s:subfield` and validation cohorts.  The
`k8s:optional` ensures that we don't run the name validation if name is
empty, because core apimachinery will already flag it as Required().

This demonstrates some of the DV value - docs and clients are now (in
theory) able to see what RC's name format is.

Co-Authored-by: Yongrui Lin <yongrlin@outlook.com>

ballista01 committed:
fix overlapping client CA and requestheader CA validation with proper certificate checking

fix error message

fix lint

thockin committed:
Fix field path for embedded fields in root types

Given the following:

```
type Struct type {
    TypeMeta int

    // +k8s:validateFalse="embedded"
    OtherStruct
}
```

What should the field-path be for the error?

or:

```
type Struct type {
    TypeMeta int

    OtherStruct // embedded
}

// +k8s:validateFalse="otherstruct"
type OtherStruct {
  I int
}
```

Conclusion: If we find an embedded field, we look for a nil fldPath, and
if so we set it to the type name.  Embedded values in root types (e.g.
above `spec`) should not be a things we actually do.

Implementation: Add `safe.Value(*T, func() *T)`

jm-franc committed:
Add configurable tolerance e2e test.

thockin committed:
Add support for validation cohorts

On a given field or type, there may be multiple "cohorts" of validation
which need to be processed together.  For example, a short-circuit
validation and a non-short-circuit for the same subfield.

The unnamed cohort (aka the default cohort) is first, followed by named
cohorts in order they are created.  Named cohorts are emitted as inline
functions, so early-return can be used.

This allows tags like k8s:subfield or k8s:item to generate cohorts
(named after the field's JSON name or the selector string,
respectively).

Subsequent commits will add support to those tags.

Co-Authored-by: Yongrui Lin <yongrlin@outlook.com>

thockin committed:
Add cohort support to +k8s:subfield

Given:

```
    // +k8s:subfield(name)=+k8s:optional
    // +k8s:subfield(name)=+k8s:format=dns-label
    // +k8s:subfield(generateName)=+k8s:optional
    // +k8s:subfield(generateName)=+k8s:format=dns-label
    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```

...we get:

```
  errs = append(errs,
    func(fldPath *field.Path, obj, oldObj *metav1.ObjectMeta) (errs field.ErrorList) {
      func() { // cohort name
        if e := validate.Subfield(ctx, op, fldPath, obj, oldObj, "name", func(o *metav1.ObjectMeta) *string { return &o.Name }, validate.OptionalValue); len (e) != 0 {
          return // do not proceed
        }
        errs = append(errs, validate.Subfield(ctx, op, fldPath, obj, oldObj, "name", func(o *metav1.ObjectMeta) *string { return &o.Name }, validate.DNSLabel)...)
      }()
      func() { // cohort generateName
        if e := validate.Subfield(ctx, op, fldPath, obj, oldObj, "generateName", func(o *metav1.ObjectMeta) *string { return &o.GenerateName }, validate.OptionalValue); len(e) != 0
 {
          return // do not proceed
        }
        errs = append(errs, validate.Subfield(ctx, op, fldPath, obj, oldObj, "generateName", func(o *metav1.ObjectMeta) *string { return &o.GenerateName }, validate.DNSLabel)...)
      }()
      return
    }(fldPath.Child("metadata"), &obj.ObjectMeta, safe.Field(oldObj, func(oldObj *corev1.ReplicationController) *metav1.ObjectMeta { return &oldObj.ObjectMeta }))...)
```

It is worth noting that this only works one level deep:

```
    // +k8s:subfield(foo)=+k8s:subfield(bar)=+k8s:optional
    // +k8s:subfield(foo)=+k8s:subfield(bar)=+k8s:format=dns-label
    Field StructType
```

...will generate one cohort ("foo") with two distinct calls to Subfield,
so the optional will not short-circuit format.  If we need deeper
cohorting we need to either make subfield take multiple field names
(e.g. `subfield(foo.bar)` or `subbfield(foo, bar)`) or we need to
accumulate "structure" in addition to validator calls.  This is similar
to how we handle multiple `eachVal` tags - each one iterates the list.

thockin committed:
Add cohort support to +k8s:item

Given:

```
   // +k8s:item(stringKey: "target", intKey: 42, boolKey: true)=+k8s:validateFalse="item Items[stringKey=target,intKey=42,boolKey=true] 1"
   // +k8s:item(stringKey: "target", intKey: 42, boolKey: true)=+k8s:validateFalse="item Items[stringKey=target,intKey=42,boolKey=true] 2"
   Items []Item `json:"items"`
```

...we get:

```
   func() { // cohort {"stringKey": "target", "intKey": 42, "boolKey": true}
       errs = append(errs, validate.SliceItem(ctx, op, fldPath, obj, oldObj, func(item *Item) bool { return item.StringKey == "target" && item.IntKey == 42 && item.BoolKey == true }, validate.DirectEqual, func(ctx context.Context, op operation.Operation, fldPath *field.Path, obj, oldObj *Item) field.ErrorList {
           return validate.FixedResult(ctx, op, fldPath, obj, oldObj, false, "item Items[stringKey=target,intKey=42,boolKey=true] 1")
       })...)
       errs = append(errs, validate.SliceItem(ctx, op, fldPath, obj, oldObj, func(item *Item) bool { return item.StringKey == "target" && item.IntKey == 42 && item.BoolKey == true }, validate.DirectEqual, func(ctx context.Context, op operation.Operation, fldPath *field.Path, obj, oldObj *Item) field.ErrorList {
           return validate.FixedResult(ctx, op, fldPath, obj, oldObj, false, "item Items[stringKey=target,intKey=42,boolKey=true] 2")
       })...)
   }()
```

thockin committed:
Add ValidateObjectMetaWithOpts() to apimachinery

This new function does 1 main thing (changes how name validation works),
and the rest is future-proofing.  Specifically this changes the
signature of the name validation function.

Before: Name validation functions return `[]string` and apimachinery
logic wraps those in `field.Invalid()`.

Problem: This makes it hard to incrementally add declarative validation,
which needs access to potential name and generateName errors to set
origins properly.

After: New name validation functions return `field.ErrorList`, which can
be wrapped in trivial functions to do things like "mark as covered by
declarative" in just one place rather than for all types.

The "WithOpts" part of this gives us a new function name which isn't
horrible and is plausibly useful in the future.

Another notable change is that this no longer directly validates the
`generateName` field (but only on the new path -- existing code still
behaves the same).  From the code:

```
+   // generateName is not directly validated here. Types can have
+   // different rules for name generation, and the nameFn is for validating
+   // the post-generation data, not the input. In the past we assumed that
+   // name generation was always "append 5 random characters", but that's not
+   // NECESSARILY true. Also, the nameFn should always be considering the max
+   // length of the name, and it doesn't know enough about the name generation
+   // to do that. Also, given a bad generateName, the user will get errors
+   // for both the generateName and name fields. We will focus validation on
+   // the name field, which should give a better UX overall.
```

This also beefs up tests a bit.

thockin committed:
Update CSR DV test to match RC style

Once we go broad, people will copy these.  Let's make them easy to debug
:)

thockin committed:
Prefactor: Fix some bad tests

thockin committed:
Make ErrorMatcher more strict about multi-match

As discussed in a proposed PR, multi-match makes sense when origin is
specified and matched, but otherwise it feels "loose" to multi-match.
And in fact it was hiding bugs in declarative validation tests and real
bugs (in subsequent commits) in implementation.

Now we allow multi-match if and only if:
  1) ErrorMatcher.ByOrigin() was specified
  2) The origin string is not empty
  3) The multiple "got" errors are not exactly identical

This last part is key -- if some change to DV codegens logic which
(errantly) calls the same function twice, we want to know about it.

thockin committed:
Remove ExpectRegexpsByPath()

All use cases are converted.

Co-Authored-by: Yongrui Lin <yongrlin@outlook.com>

thockin committed:
Remove ExpectInvalid()

All use cases are converted.

Co-Authored-by: Yongrui Lin <yongrlin@outlook.com>

thockin committed:
Add comments

thockin committed:
Fix ReplicationControl double validation

BenTheElder committed:
check invariant metrics after e2e tests

richabanker committed:
Wrapper for DeleteLabelValues

eltrufas committed:
Mark API server errors as transient in csi raw block driver

Certain failures during SetupDevice and MapPodDevice are not treated as
transient in the csi raw block plugin implementation, while they are in
the file mode plugin. This can lead to certain failures causing volumes
to be marked as unmounted incorrectly.

This patch brings the block plugin up to parity with the fs one by
marking the equivalent calls as transient. This mostly covers API server
and some csi driver calls.

tallclair committed:
Introduce API to codify and validate feature gate dependencies

pohly committed:
DRA: CRUD conformance tests

Promoting real tests turned out to be harder than expected (should be rewritten
to be self-contained, additional reviews, etc.).

They would not achieve 100% endpoint+operation coverage because real tests only
use some of the operations. Therefore each API type has to be covered with
CRUD-style tests which only exercise the apiserver, then maybe additional
functional tests can be added later (depending on time and motivation).

The machinery for testing different API types is meant to be reusable, so it
gets added in the new e2e/framework/conformance helper package.

natasha41575 committed:
separate resource-quota and limit-ranger resize tests

natasha41575 committed:
expand coverage for resource quota and limit ranger tests

pohly committed:
DRA E2E: only check ResourceClaims in the test namespace

That was the original intent, but the implementation then ended up checking
ResourceClaims in all namespaces. Depending on timing this was merely
misleading (showing ResourceClaim changes from a different test running in
parallel), but with upcoming CRUD tests which intentionally set an allocation
result without a finalizer it breaks the non-CRUD tests when they check the
those CRUD ResourceClaims.

pohly committed:
build: also support KUBE_RACE for test binaries

This is relevant for building e2e.test in
pull-kubernetes-e2e-kind-alpha-beta-features-race: if it contains data races,
tests might be flaky.

kei01234kei committed:
make v1 resource version first priority in resource

kei01234kei committed:
modify api version hash

kei01234kei committed:
modify etcd data for integration test

lmktfy committed:
Fix OWNERS_ALIASES entry for SIG Docs

pohly committed:
DRA scheduler: clean up feature gate handling

Copying from feature.Features to new fields in the plugin got a bit silly with
the long list of features that we have now. Embedding feature.Features is
simpler.

Two fields in feature.Features weren't named according to the feature gate, now
they are named consistently and the fields are sorted.

pohly committed:
DRA scheduler: add unit test for allocator selection

This prevents the mistake from 1.34 where the default-on
DRAResourceClaimDeviceStatus feature caused the use of the experimental
allocator implementation. The test fails without a fix for that.

pohly committed:
DRA scheduler: fix selection of "incubating" allocator implementation

In 1.34, the default feature gate selection picked the "experimental" allocator
implementation when it should have used the "incubating" allocator. No harm
came from that because the experimental allocator has all the necessary if
checks to disable the extra code and no bugs were introduced when implementing
it, but it means that our safety net wasn't there when we expected it to be.

The reason is that the "DRAResourceClaimDeviceStatus" feature gate is on by
default and was only listed as supported by the experimental implementation.
This could be fixed by listing it as supported also by the other
implementation, but that would be a bit odd because there is nothing to support
for it (the reason why this was missed in 1.34!). Instead, the allocator
features are now only indirectly related to feature gates, with a single
boolean controlling the implementation of binding conditions.

pohly committed:
DRA scheduler_perf: run with specific allocator implementations

For some packages (in particular the code DRA), all allocator implementations
can handle the testcases. Some other packages are for less stable features and
work with fewer implementations. Now all unit tests are run with all suitable
implementations, to increase code coverage.

Benchmarks are fixed to the most mature implementation because they would be
costly to run in more than one. When promoting an allocation implementation we
can do before/after comparisons to detect potential performance regressions.

The downside of this approach is that we need to remember to extend the list
of supported implementations when promoting features, otherwise testing will miss
some new supported implementation.

pohly committed:
scheduler_perf: apply feature gates in deterministic, alphabetical order

Without this, the effect of the following feature gate config would be random:

    featureGates:
      AllBeta: false
      SomeBetaFeature: true

That's random because the order of iterating of the map is randomized by Go and
`AllBeta: false` would disable `SomeBetaFeature` if (and only if) applied last.

Now by sorting alphabetically, AllAlpha/Beta come first in practice. It's not a
complete solution, some future feature gate name might come before it.

pohly committed:
build: automatically choose a suitable base image

When building some command dynamically through KUBE_CGO_OVERRIDES, the base
image for that command must provide a libc which matches what the command was
compiled against. If the runtime libc is too old, then it might not have the
right ABI (missing symbols or wrong version of those symbols).

A user or job shouldn't need to know which of the different Kubernetes images
is the right one, therefore the build scripts now use the same image that was
previously only used for kube-proxy also for other commands if they get build
dynamically. This does not increase the maintenance effort because a single
image bump remains sufficient and for e.g. a kind cluster it doesn't matter
that the kube-proxy base image is slightly larger than necessary for the other
commands - it only needs to be present once in the node image.

AwesomePatrol committed:
Move metrics calculations to getList

Signed-off-by: Aleksander Mistewicz <amistewicz@google.com>

AwesomePatrol committed:
Move stats updates to getList

Signed-off-by: Aleksander Mistewicz <amistewicz@google.com>

huww98 committed:
scheduler/volumebinding: use subtest

huww98 committed:
scheduler/volumebinding: pass testing.T to helper

huww98 committed:
scheduler/volumebinding: introduce testInformer

huww98 committed:
scheduler/volumebinding: passive assume cache

Currently volume and dynamic-resource plugin shares an AssumeCache
implementation. However, they features significantly different use case.  DRA
call Assume() on objects returned by APIServer, but volume call Assume() on
objects yet to be sent to the APIServer.

VolumeBinding plugin only makes one update request, while DynamicResource makes
2 requests (add finalizer then update allocation status).  Taking advantage of
this, currently the volume cache is simpler:

1. Reserve: assume PV/PVC will be updated
2. PreBind: really send the update request
3. AssumeCache receives an update from informer and overwrite the assumed state.
   a. if Prebind succeeded, this will surely include the update from step 2.
   b. if PreBind is not finished yet, and this is an irrelevant update, it is safe to
      overwrite the assumed state, because our update in PreBind will surely fail with Conflict.

While for DynamicResource:

1. Reserve: add devices to inFlightAllocations
2. PreBind:
   a. send the 2 update requests
   b. add the returned object into AssumeCache
   c. AssumeCache dispatch events synchronously to update allocatedDevices
   d. remove devices from inFlightAllocations

DynamicResource needs some features from AssumeCache that is not necessary for VolumeBinding:
1. DynamicResource needs strictly ordered update events to update allocatedDevices,
   including those from Assume() and Restore()
2. DynamicResource needs to compare ResourceVersion to prevent the assumed state from being
   overwritten by older version from informer.  While this works, the doc[1] says:
   "you must not compare resource versions for greater-than or less-than relationships".

Given so much difference, it can be beneficial to fork another simpler
AssumeCache for VolumeBinding plugin. Because of no need to send events, the
lite AssumeCache is a passive component. It only record the assumed version
without copying all objects from informer into its local cache. When reading,
we read from both informer and local cache.  So it will always be up-to-date
with informer, no need to wait for event handler.

This resolves a race condition where AssumeCache and scheduler queue both
receive events from informer. When a pod is scheduling due to PV update event,
the PVCache may be not updated yet because it has not processed the relevant
event.

The passive version still listens events from informer, but only for cleaning up
its local cache to save memory.

[1]: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions

huww98 committed:
scheduler/volumebinding: remove Get[API]{PV,PVC}

should be replaced by generic Get[APIObj]

Jefftree committed:
Rename CLE test directories

huww98 committed:
scheduler/volumebinding: target AssumeCache UT generic passiveAssumeCache

And removing duplicate tests

huww98 committed:
scheduler/volumebinding: rename passive_assume_cache_test.go

huww98 committed:
scheduler/volumebinding: add test for PVAssumeCache

aaron-prindle committed:
feat: make it so that all ShortCircuit validators are run for short circuiting check

rutigs committed:
Record SyncTerminatingPod error and set error status on span when non-nil err returned

Signed-off-by: Nick Rutigliano <rutigliano.nick@gmail.com>

pohly committed:
E2E: treat data races in e2e suite as failures

Ginkgo itself doesn't do this, in which case prun-junit-xml drops the output
and Spyglass wouldn't show the test as failed. If the data race warning is
captured, we now treat that as the failure of a test if it hasn't already
failed for other reasons.

While at it, the entire report cleanup gets moved to our junit package.

pohly committed:
test: automatically lower Ginkgo parallelism when using race detection

When race detection is enabled, merely running 25 e2e.test instances was too much
and the OOM killer shut down the Prow test pod because of the memory overhead.

A CI job could control that via GINKGO_PARALLEL_NODES, but we should also have
saner defaults which take this into account.

aojea committed:
integration test: webhook proxy behavior

adds a new integration test to verify that the API server's egress
to admission webhooks correctly respects the standard `HTTPS_PROXY`
and `NO_PROXY` environment variables.

It adds a new test util to implement a Fake DNS server that allows
to override DNS resolution in tests, specially useful for integration
test that can only bind to localhost the servers, that is ignored
by certain functionalities.

aojea committed:
improve httpstream handshake error logging

Some tests are flaking in this function, having a better logged
error will help to understand the cause.

Also modify the Handshake function to always write an http error,
otherwise we have inconsistencies in the codebase that is using the
function and since we are already passing the http writer it makes sense
to not just return an error and delegate the response to the caller.

jpbetz committed:
Add maxItems limits to ResourceClaim

jpbetz committed:
Add declarative validation tests, use tweak pattern, and additional test structure changes

jpbetz committed:
simplify scale subresource testing and document expectations

Co-Authored-by: Yongrui Lin <yongrlin@outlook.com>

swagatbora90 committed:
change klog to log volume name instead of entire spec in GetDeviceMountPath

Signed-off-by: Swagat Bora <sbora@amazon.com>

yongruilin committed:
Refactor: Centralize declarative validation and migration logic

The boilerplate for running declarative validation was duplicated across multiple resource strategies. This included feature gate checks, metric identifier generation, error comparison, and conditional merging logic, which made the code verbose and difficult to maintain.

This commit introduces a new helper function, `rest.ValidateDeclarativelyWithMigrationChecks`, to encapsulate this common logic. All relevant strategies have been refactored to use this new function, resulting in cleaner and more concise code.

yongruilin committed:
test: Add unit tests for metricIdentifier function

aditigupta96 committed:
refactor(apiextensions-apiserver): Make NamingConditionController fully context-aware

yongruilin committed:
refactor: Remove Validate(Update)Declaratively and improve error handling

yongruilin committed:
Remove unused WithTakeover and WithValidationIdentifier

toVersus committed:
deflake e2e: ensure pod with sidecars restarts in correct order after node reboot

Signed-off-by: Tsubasa Nagasawa <toversus2357@gmail.com>
Co-authored-by: Hironori Shiina <shiina.hironori@gmail.com>

aaron-prindle committed:
feat: migrate k8s:format=k8s-long-name-caseless validation on OpaqueDeviceConfiguration.Driver to declarative validation

DanBokete committed:
Deprecate caseless driver name validation and enforce lowercase warnings

- Deprecate IsDNS1123SubdomainCaseless to avoid caseless validation issues.
- Warn when ResourceSlice driver names contain uppercase characters.
- Clarify driver names must be DNS subdomains and use only lowercase letters.
- Update tests, staging code, and OpenAPI spec to reflect the changes.

hoskeri committed:
Fix IPv6 allocator for /64 CIDRs

An ipAllocator with a 64 bit IPv6 CIDR can allocate
addresses outside the CIDR range, due to an improper uint64 to int64
cast in the addOffsetAddress function.

Replace the cast with a call to `math/big.Int.SetUint64()`.

danwinship committed:
Update comments around an old bug workaround

n2h9 committed:
[126379] [go-client] chore: use WithContext functions

Signed-off-by: Nikita B <n2h9z4@gmail.com>

aaron-prindle committed:
add +k8s:maxItems tag logic and tests

aaron-prindle committed:
generate

aaron-prindle committed:
fix validation_resourceclaim_test.go with MarkCoveredByDeclarative

macsko committed:
Disable SchedulerAsyncAPICalls feature gate due to a known regression

natasha41575 committed:
refactor and increase coverage for pod resize tests

aaron-prindle committed:
refactor(DRA validation): Add granular controls to ValidateCSIDriverName for declarative validation migration

aditigupta96 committed:
refactor(apiextensions-apiserver): Make NonStructuralSchema controller context-aware

n2h9 committed:
[126379] [go-client] chore: use WithContext functions: do not use SleepWithContext inside Sleep, use CalculateBackoff inside CalculateBackoffWithContext

Signed-off-by: Nikita B <n2h9z4@gmail.com>

joshjms committed:
bump etcd to v3.6.5

Signed-off-by: joshjms <joshjms1607@gmail.com>

liggitt committed:
Drop runtime use of gogo packages

liggitt committed:
Regenerate protobuf

liggitt committed:
Regenerate vendor

aaron-prindle committed:
fix: add +enum tag to resource DeviceAllocatoionMode

mayank-agrwl committed:
Add RunWithContext method for debugsocket

mayank-agrwl committed:
Make APIServerLeaseGC controller context-aware

mayank-agrwl committed:
Make legacytokentracking controller context aware

raywainman committed:
update autoscaling leads

p0lyn0mial committed:
client-go/util/watchlist/watch_list: intro DoesClientNotSupportWatchListSemantics

aaron-prindle committed:
fix: update pkg/apis/resource validateDeviceRequest to short-circuit on the oneOf check

cfryanr committed:
add getters for event User and ImpersonatedUser on AuditContext

The ability to get these values was deleted by #129472 which removed
the public function audit.AuditEventFrom. The new getters return copies
of the underlying objects to prevent external mutations of the objects
held by the audit events.

itzPranshul committed:
Enable validation-gen on rbac/v1,v1alpha1,v1beta1

This enables generation but doesn't activate the strategy calling of declarative

itzPranshul committed:
clusterRole:Enable declarative validation

Add declarative validation to Validate and ValidateUpdate

ania-borowiec committed:
Fix in code and integration test that verifies that when victim pod is stuck in binding, preemptor pod remains waiting in unschedulable queue until deletion of the victim pod is completed

omerap12 committed:
Add desired_replicas histogram metric to HPA controller

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

itzPranshul committed:
tests: add declarative validation migration tests for ClusterRole

Introduce a test suite that ensures declarative test cases
are fully exercised and that validation errors are compared
with handwritten (imperative) validation to ensure consistency.

ialidzhikov committed:
Remove configmaps related rules from the kube-controller-manager and kube-scheduler leader election roles

yongruilin committed:
chore: Move declarative validation featuregates to staging apiserver

lalitc375 committed:
feat(resource): Add maxItems validation for DeviceClass selectors

This commit introduces a validation rule to limit the number of selectors
in a DeviceClass to 32. This is done by adding the `+k8s:maxItems=32`
marker to the DeviceClassSpec and running the code generator.

The following changes are included:
- Updated `staging/src/k8s.io/api/resource/v*/types.go` with the validation marker.
- Regenerated validation code in `pkg/apis/resource/v*/zz_generated.validations.go`.
- Improved error message in `pkg/apis/resource/validation/validation.go`.
- Added tests in `pkg/registry/resource/deviceclass/declarative_validation_test.go`
  to cover the new validation.

lalitc375 committed:
feat(resource): Add maxItems validation for DeviceClass configs

humblec committed:
pkg/volume: improve code style and readability

- Improve slice declaration formatting in quota_common_linux_impl.go
- Enhance variable naming clarity (myMPSlash -> myMountPointWithSlash)
- Standardize function comment format for consistency
- Add trailing comma to slice declarations following Go conventions

These are non-breaking  improvements that enhance code
readability and maintain consistency with Go best practices.

Signed-off-by: Humble Devassy Chirammal <humble.devassy@gmail.com>

humblec committed:
test: update CSI sidecar images to latest versions

Update CSI sidecar container images in test manifests:
- livenessprobe: v2.15.0 → v2.17.0
- csi-attacher: v4.8.0 → v4.10.0
- csi-provisioner: v5.1.0 → v5.3.0
- csi-resizer: v1.13.1 → v1.14.0

Signed-off-by: Humble Devassy Chirammal <humble.devassy@gmail.com>

chizhang21 committed:
fix(cordonhelper): Avoid mutating local node before API call

Signed-off-by: cz21 <cz21ok@gmail.com>

humblec committed:
cluster/addons: update container images to latest versions

Update several addon container images to their latest stable versions:
- metrics-server: v0.7.2 → v0.8.0
- addon-resizer: 1.8.14 → 1.8.20
- ip-masq-agent: v2.9.3 → v2.9.4
- kube-network-policies: v0.7.0 → v0.8.0

These updates provide bug fixes, security improvements, and enhanced functionality

Signed-off-by: Humble Devassy Chirammal <humble.devassy@gmail.com>

alaypatel07 committed:
fix resource claims deallocation for extended resource when pod is completed

Signed-off-by: Alay Patel <alayp@nvidia.com>

xigang committed:
nodelifecycle: fix ComputeZoneState method comment

Signed-off-by: xigang <wangxigang2014@gmail.com>

xigang committed:
disruption: remove unused pdb parameter from getExpectedScale method

Signed-off-by: xigang <wangxigang2014@gmail.com>

neolit123 committed:
kubeadm: rework the FetchInitConfigurationFromCluster node flags

The newControlPlane flag has been historically problematic, since
it implies that the function FetchInitConfigurationFromCluster
cannot handle the cases where a node is worker node but
we still want to fetch its NodeRegistrationOptions conditionally,
in cases such as "upgrade node" for workers.

To fix this issue, replace the flag newControlPlaneNode with
two new flags getNodeRegistration and getAPIEndpoint.

If getNodeRegistration is true, we fetch the NRO, and if
getAPIEndpoint is true, we fetch the API endpoint for
that node.

Additionally, rename skipComponentConfigs to getComponentConfigs
for consistency and flip its value accordingly everywhere.

yongruilin commented on pkg/registry/rbac/clusterrole/declarative_validation_test.go:
```diff
@@ -0,0 +1,137 @@
+		t.Run(k, func(t *testing.T) {
+			var declarativeTakeoverErrs field.ErrorList
+			var imperativeErrs field.ErrorList
+			for _, gateVal := range []bool{true, false} {
+				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidation, gateVal)
+				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidationTakeover, gateVal)
+
+				errs := Strategy.Validate(ctx, &tc.input)
+				if gateVal {
+					declarativeTakeoverErrs = errs
+				} else {
+					imperativeErrs = errs
+				}
+				errOutputMatcher := field.ErrorMatcher{}.ByType().ByField().ByOrigin()
+				if len(tc.expectedErrs) > 0 {
+					errOutputMatcher.Test(t, tc.expectedErrs, errs)
+				} else if len(errs) != 0 {
+					t.Errorf("expected no errors, but got: %v", errs)
+				}
+			}
+			equivalenceMatcher := field.ErrorMatcher{}.ByType().ByField().ByOrigin()
+			equivalenceMatcher.Test(t, imperativeErrs, declarativeTakeoverErrs)
+			apitesting.VerifyVersionedValidationEquivalence(t, &tc.input, nil)
```
Could you replace this part with https://github.com/kubernetes/kubernetes/blob/5161bf00587eb36cd3ce64635126f47a98068b5c/pkg/api/testing/validation.go#L172
You can refer to the usage at https://github.com/kubernetes/kubernetes/blob/5161bf00587eb36cd3ce64635126f47a98068b5c/pkg/registry/resource/resourceclaim/declarative_validation_test.go#L61

yongruilin commented on pkg/registry/rbac/clusterrole/declarative_validation_test.go:
```diff
@@ -0,0 +1,137 @@
+			var declarativeTakeoverErrs field.ErrorList
+			var imperativeErrs field.ErrorList
+			for _, gateVal := range []bool{true, false} {
+				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidation, gateVal)
+				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.DeclarativeValidationTakeover, gateVal)
+				errs := Strategy.ValidateUpdate(ctx, &tc.update, &tc.old)
+				if gateVal {
+					declarativeTakeoverErrs = errs
+				} else {
+					imperativeErrs = errs
+				}
+				errOutputMatcher := field.ErrorMatcher{}.ByType().ByField().ByOrigin()
+
+				if len(tc.expectedErrs) > 0 {
+					errOutputMatcher.Test(t, tc.expectedErrs, errs)
+				} else if len(errs) != 0 {
+					t.Errorf("expected no errors, but got: %v", errs)
+				}
+			}
+			equivalenceMatcher := field.ErrorMatcher{}.ByType().ByField().ByOrigin()
+			dedupedImperativeErrs := field.ErrorList{}
+			for _, err := range imperativeErrs {
+				found := false
+				for _, existingErr := range dedupedImperativeErrs {
+					if equivalenceMatcher.Matches(existingErr, err) {
+						found = true
+						break
+					}
+				}
+				if !found {
+					dedupedImperativeErrs = append(dedupedImperativeErrs, err)
+				}
+			}
+			equivalenceMatcher.Test(t, dedupedImperativeErrs, declarativeTakeoverErrs)
+			apitesting.VerifyVersionedValidationEquivalence(t, &tc.update, &tc.old)
```
Ditto. Could you replace it with the test helper?

itzPranshul committed:
Merge branch 'kubernetes:master' into master-vg-enable-clusterRole

itzPranshul committed:
Fix ClusterRole declarative validation tests and strategy DV migration

itzPranshul committed:
Formatting: Run update-gofmt.sh

p0lyn0mial committed:
test/e2e/apimachinery/watchlist: select only wellknown secrets for table test

pohly committed:
DRA ResourceSlice tracker: explain test a bit better, fix -run

Instead of creating a new test case, the permutation is passed down. This
enables adding the event numbers to the log output, which is useful to
understand better which output belongs to which input:

    === RUN   TestListPatchedResourceSlices/update-patch/2_3_0_1
    tracker.go:396: I0929 14:28:40.032318] event #1: ResourceSlice add slice="s1"
    tracker.go:581: I0929 14:28:40.032404] event #1: syncing ResourceSlice resourceslice="s1"
    tracker.go:659: I0929 14:28:40.032446] event #1: ResourceSlice synced resourceslice="s1" change="add"
    tracker.go:396: I0929 14:28:40.032502] event #2: ResourceSlice add slice="s2"
    tracker.go:581: I0929 14:28:40.032536] event #2: syncing ResourceSlice resourceslice="s2"
    tracker.go:659: I0929 14:28:40.032568] event #2: ResourceSlice synced resourceslice="s2" change="add"
    tracker.go:463: I0929 14:28:40.032609] event #0/#0: DeviceTaintRule add patch="rule"
    tracker.go:581: I0929 14:28:40.032639] event #0/#0: syncing ResourceSlice resourceslice="s1"
    tracker.go:703: I0929 14:28:40.032675] event #0/#0: processing DeviceTaintRule resourceslice="s1" deviceTaintRule="rule"
    tracker.go:807: I0929 14:28:40.032712] event #0/#0: applying matching DeviceTaintRule resourceslice="s1" deviceTaintRule="rule" device="driver1.example.com/pool-1/device-1"
    tracker.go:868: I0929 14:28:40.032780] event #0/#0: Assigned new taint ID, no matching taint resourceslice="s1" deviceTaintRule="rule" device="driver1.example.com/pool-1/device-1" taintID=0 taint="example.com/taint=tainted:NoExecute"
    tracker.go:654: I0929 14:28:40.033023] event #0/#0: ResourceSlice synced resourceslice="s1" change="update" diff=<
        	@@ -23,7 +23,32 @@
        	     "BindingConditions": null,
        	     "BindingFailureConditions": null,
        	     "AllowMultipleAllocations": null,
        	-    "Taints": null
        	+    "Taints": [
        	+     {
        	+      "Rule": {
        	+       "metadata": {
        	+        "name": "rule"
        	+       },
        	+       "spec": {
        	+        "deviceSelector": {
        	+         "pool": "pool-1"
        	+        },
        	+        "taint": {
        	+         "key": "example.com/taint",
        	+         "value": "tainted",
        	+         "effect": "NoExecute",
        	+         "timeAdded": "2006-01-02T15:04:05Z"
        	+        }
        	+       },
        	+       "status": {}
        	+      },
        	+      "ID": 1,
        	+      "key": "example.com/taint",
        	+      "value": "tainted",
        	+      "effect": "NoExecute",
        	+      "timeAdded": "2006-01-02T15:04:05Z"
        	+     }
        	+    ]
        	    }
        	   ],
        	   "Taints": null,
         >
    tracker.go:482: I0929 14:28:40.033224] event #0/#1: DeviceTaintRule update patch="rule" diff=<
        	@@ -4,7 +4,7 @@
        	  },
        	  "spec": {
        	   "deviceSelector": {
        	-   "pool": "pool-1"
        	+   "pool": "pool-2"
        	   },
        	   "taint": {
        	    "key": "example.com/taint",
         >
    tracker.go:581: I0929 14:28:40.033285] event #0/#1: syncing ResourceSlice resourceslice="s1"
    tracker.go:703: I0929 14:28:40.033319] event #0/#1: processing DeviceTaintRule resourceslice="s1" deviceTaintRule="rule"
    tracker.go:654: I0929 14:28:40.033478] event #0/#1: ResourceSlice synced resourceslice="s1" change="update" diff=<
        	@@ -23,32 +23,7 @@
        	     "BindingConditions": null,
        	     "BindingFailureConditions": null,
        	     "AllowMultipleAllocations": null,
        	-    "Taints": [
        	-     {
        	-      "Rule": {
        	-       "metadata": {
        	-        "name": "rule"
        	-       },
        	-       "spec": {
        	-        "deviceSelector": {
        	-         "pool": "pool-1"
        	-        },
        	-        "taint": {
        	-         "key": "example.com/taint",
        	-         "value": "tainted",
        	-         "effect": "NoExecute",
        	-         "timeAdded": "2006-01-02T15:04:05Z"
        	-        }
        	-       },
        	-       "status": {}
        	-      },
        	-      "ID": 1,
        	-      "key": "example.com/taint",
        	-      "value": "tainted",
        	-      "effect": "NoExecute",
        	-      "timeAdded": "2006-01-02T15:04:05Z"
        	-     }
        	-    ]
        	+    "Taints": null
        	    }
        	   ],
        	   "Taints": null,
         >
    tracker.go:581: I0929 14:28:40.033601] event #0/#1: syncing ResourceSlice resourceslice="s2"
    tracker.go:703: I0929 14:28:40.033633] event #0/#1: processing DeviceTaintRule resourceslice="s2" deviceTaintRule="rule"
    ...

Disabling event checking only worked when actually running all sub-tests. When
selectively running only one permutation with -run, the boolean variable was
wrong:

    $ go test -run='.*/^update-patch$' ./staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/
    ok  	k8s.io/dynamic-resource-allocation/resourceslice/tracker

    $ go test -run='.*/^update-patch$/3_2_0_1' ./staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/
    --- FAIL: TestListPatchedResourceSlices (0.01s)
        --- FAIL: TestListPatchedResourceSlices/update-patch (0.00s)
            --- FAIL: TestListPatchedResourceSlices/update-patch/3_2_0_1 (0.00s)

                tracker_test.go:762:
                     	Error Trace:	/nvme/gopath/src/k8s.io/kubernetes/staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker_test.go:762
                     	            				/nvme/gopath/src/k8s.io/kubernetes/staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker/tracker_test.go:856
                    	Error:      	Not equal:
                    	            	expected: []tracker.handlerEvent{tracker.handlerEvent{event:"add", oldObj:(*api.ResourceSlice)(nil), newObj:(*api.ResourceSlice)(0xc000301d40)}, tracker.handlerEvent{event:"add", oldObj:(*api.ResourceSlice)(nil), newObj:(*api.ResourceSlice)(0xc000346000)}}
                    	            	actual  : []tracker.handlerEvent{tracker.handlerEvent{event:"add", oldObj:(*api.ResourceSlice)(nil), newObj:(*api.ResourceSlice)(0xc0001f9ba0)}, tracker.handlerEvent{event:"add", oldObj:(*api.ResourceSlice)(nil), newObj:(*api.ResourceSlice)(0xc000301d40)}, tracker.handlerEvent{event:"update", oldObj:(*api.ResourceSlice)(0xc000301d40), newObj:(*api.ResourceSlice)(0xc0003dba00)}, tracker.handlerEvent{event:"update", oldObj:(*api.ResourceSlice)(0xc0003dba00), newObj:(*api.ResourceSlice)(0xc000301d40)}, tracker.handlerEvent{event:"update", oldObj:(*api.ResourceSlice)(0xc0001f9ba0), newObj:(*api.ResourceSlice)(0xc0003dbba0)}}

Now permutations are detected automatically based on the indices.

While at it, documentation gets moved around a bit to make reading test cases
easier without going to the implementation.

pohly committed:
DRA ResourceSlice: nicer log output

UniqueString rendered as {} in JSON/YAML, which isn't useful... Providing a
MarshalJSON which dumps the string is better.

This enables full object dumps in the ResourceSlice tracker. While at it:
- Consistently log the object content at level 6.
- Use past tense in log messages (recommended style).
- Rename "patch" -> "rule".

michaelasp committed:
feat: Add helper function for client-go to compare resource version

lalitc375 committed:
Add listType=map and listMapekey=uid for ResourceClaimStatus.ReservedFor

michaelasp committed:
feat: Add conformance tests for all resources for comparable resource version

pohly committed:
DRA device taints: fix toleration of NoExecute

As usual, consumers of an allocated claim react to the information stored in
the status. In this case, the scheduler did not copy the tolerations into the
status and as a result a pod with a toleration for NoExecute got scheduled and
then immediately evicted.

Some additional logging gets added to make the handling easier to track in the
eviction controller. Example YAMLs allow reproducing the use case manually.

michaelasp committed:
Add additional types for resource version comparison testing

michaelasp committed:
feat: Add matcher and conformance tests ensuring that RV is uint128

michaelasp committed:
Document 0 as a special case in RV comparison

yongruilin committed:
Simplify tests wrt ratcheting

Co-Authored-by: Tim Hockin <thockin@google.com>

liggitt committed:
Fix passing runtime.Object to HaveValidResourceVersion check

liggitt committed:
Isolate generated ProtoMessage() methods in build-tagged files

yongruilin committed:
fix: Adjust validation for pool names to ensure proper coverage in device requests

ajaysundark committed:
Fix: Correct timestamp for node swap metrics

The node swap metric was incorrectly using the memory timestamp instead of the swap timestamp. This change corrects the timestamp and adds a test case to verify the fix.

yongruilin committed:
Revert "Omit type names of emitted slice elements to appease gofmt"

This reverts commit 243f47f3b3314e2d69d93e8c5b09974a73216378.

Not needed anymore, now that gengo calls 'gofmt -s'

yongruilin committed:
run update-codegen to for ReplicationController

yongruilin committed:
fix: Update error origin in ValidateDNS1123Label to use k8s-short-name format

yongruilin committed:
fix: Comment out ipSloppyValidator

It is currently not supported.

yongruilin committed:
fix typo in comment for namespace validation to appease verify-spelling

lalitc375 committed:
apply feedback

aaron-prindle committed:
chore: add +k8s:immutable tag implementation and test

aaron-prindle committed:
feat: add +k8s:immutable tag to ResourceClaim.spec and associated tests

aaron-prindle committed:
feat: add path normalization support to error matcher

rbiamru committed:
test/e2e_node: mark MirrorPod update tests as [NodeConformance]

humblec committed:
Update cluster/addons/kube-network-policies/kube-network-policies.yaml

Co-authored-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

humblec committed:
Update cluster/addons/ip-masq-agent/ip-masq-agent.yaml

Co-authored-by: Antonio Ojea <antonio.ojea.garcia@gmail.com>

bertinatto committed:
Update hostpathplugin image version in DRA test driver manifest

Update image from v1.7.3 to v1.16.1 to avoid pulling multiple
in CI.

pohly committed:
test: format unstructured as YAML

It's a nested map which looks a lot nicer as YAML, in particular
when it represents a Kubernetes object.

Unit+integration tests using ktesting+gomega and E2E tests benefit from this
change.

liggitt committed:
Lock all mutable fields when printing gc node

upodroid committed:
upgrade to cos 121

tchap committed:
kubectl config set-context: Add -n flag

This is simply a shorthand for --namespace as with other commands.

lalitc375 committed:
Fix rebase issues

adrianmoisey committed:
Use a dedicated delete channel in HPA test

This is an attempt to fix a flake

lalitc375 committed:
Add DV for uids in resourceClaim Status

thockin committed:
Warn when quota request > limit

BenTheElder committed:
add test to verify invariants are fully specified

BenTheElder committed:
add apiserver_validation_declarative_validation_panic_total invariant

BenTheElder committed:
add TODO to migrate when new metrics mature and old metrics are deprecated

liggitt committed:
Use lock-free for unguarded isHidden/isDeprecated fields

aaron-prindle committed:
address feedback

yongruilin committed:
feat(dra): Add maxItems validation tags for ResourceClaim fields

This commit adds +k8s:maxItems tags to several slice fields within the ResourceClaimSpec across versions v1, v1beta1, and v1beta2.

The following fields under ResourceClaimSpec are now validated for their maximum number of items:
 - DeviceRequest.FirstAvailable
 - ExactDeviceRequest.Selectors
 - DeviceSubRequest.Selectors
 - DeviceConstraint.Requests
 - DeviceClaimConfiguration.Requests

yongruilin committed:
run update-codegen.sh

humblec committed:
correct metrics function comments

Signed-off-by: Humble Devassy Chirammal <humble.devassy@gmail.com>

aojea committed:
add integration test for endpointslice controller headless services

macsko committed:
Update SIG Scheduling leads in feature-approvers

aaron-prindle committed:
address feedback

yongruilin committed:
feat(validation): Enhance validation tests with normalization rules support

yongruilin committed:
feat(validation): Add normalization rules for ResourceClaim validation

SergeyKanzhelev committed:
remove cpuSharesToCPUWeight that is not used any longer

SergeyKanzhelev committed:
reordered changelog for cri api

SergeyKanzhelev committed:
changelog for 1.34

aaron-prindle committed:
fix: remove unnecessary ValidateResourceClaim call from resourclaim ValidateUpdate

aaron-prindle committed:
generate

HirazawaUi committed:
Fix incorrect error messages

pohly committed:
E2E volume: fix restarting of watch

Presumably
https://github.com/kubernetes/kubernetes/pull/127260/files#r2405215911
was meant to continue polling after a watch was closed by the apiserver.
This is something that can happen under load.

However, returning the error has the effect that polling stops.
This can be seen as test failures when testing with race
detection enabled:

    persistent_volumes_test.go:1101: Failed to wait for all claims to be bound: watch closed

yongruilin committed:
fix(validation): Mark covered errors in TestValidateClaim for better clarity

yongruilin committed:
Address feedback: simpliy errormatcher instantiation

aaron-prindle committed:
address feedback

aditigupta96 committed:
refactor(apiextensions-apiserver): Make APIApproval controller context-aware

gnufied committed:
With new changes we will also have a VolumeModifying condition

Fix e2e tests to take that into account

pohly committed:
E2E framework: enable data race detection only if needed

When building the test binary without race detection, we don't
need the post-processing of the JUnit file because it cannot
contain data race reports. This can be done via build tags.

k8s-release-robot committed:
CHANGELOG: Update directory for v1.35.0-alpha.1 release

soltysh committed:
Update --chunk-size flag, dropping the beta information

The --chunk-size flag has been around since October 2017
(ihttps://github.com/kubernetes/kubernetes/commit/4780ad02978e7aea80f3b50ec20c9abeec13ce69)
so it's about time we make it official.

Signed-off-by: Maciej Szulik <soltysh@gmail.com>

rkolchmeyer committed:
cluster/gce/gci: revert PATH change when using custom runc

Commit 6031ff29c144fe16e98f6fcb1d7201634f9a173e
("make containerd download more robust") updated the containerd systemd
unit to use `Environment=PATH=/home/containerd/bin:$PATH`. However,
environment variables are not evaluated in systemd units in that way.
This resulted in containerd running with the literal value
`/home/containerd/bin:$PATH` for its PATH, breaking containerd (it could
not access the apparmor_parser program).

The old way passes the PATH that configure.sh runs with to containerd,
which is not optimal, but at least it doesn't break the tests. Let's
revert to the old way.

upodroid committed:
bump cos-gpu-installer to support cos 121

BenTheElder committed:
speed up make clean by narrowing recursive chmod

itzPranshul committed:
refactor Validate and ValidateUpdate to use declarative validation helper
